pypeit/specobjs.py

```diff
@@ -134,13 +138,104 @@ def nobj(self):
+    def unpack_object(self, ret_flam=False):
+        """
+        Utility function to unpack the sobjs for one object and
+        return various numpy arrays describing the spectrum and meta
+        data. The user needs to already have trimmed the Specobjs to
+        the relevant indices for the object.
+
+        Args:
+           ret_flam (:obj:`bool`, optional):
+              If True return the FLAM, otherwise return COUNTS.
+
+        Returns:
+            tuple: Returns the following where all numpy arrays
+            returned have shape (nspec, norders) for Echelle data and
+            (nspec,) for Multislit data.
+
+                - wave (`numpy.ndarray`_): Wavelength grids
+                - flux (`numpy.ndarray`_): Flambda or counts
+                - flux_ivar (`numpy.ndarray`_): Inverse variance (of
+                  Flambda or counts)
+                - flux_gpm (`numpy.ndarray`_): Good pixel mask.
+                  True=Good
+                - meta_spec (dict:) Dictionary containing meta data.
+                  The keys are defined by
+                  spectrograph.header_cards_from_spec()
+                - header (astropy.io.header object): header from
+                  spec1d file
+    """
+
+        # Read in the spec1d file
+        norddet = self.nobj
+        if ret_flam:
+            # TODO Should nspec be an attribute of specobj?
+            nspec = self[0].OPT_FLAM.size
+        else:
+            nspec = self[0].OPT_COUNTS.size
+        # Allocate arrays and unpack spectrum
+        wave = np.zeros((nspec, norddet))
+        flux = np.zeros((nspec, norddet))
+        flux_ivar = np.zeros((nspec, norddet))
+        flux_gpm = np.zeros((nspec, norddet), dtype=bool)
+        detector = np.zeros(norddet, dtype=int)
+        ech_orders = np.zeros(norddet, dtype=int)
+        for iorddet in range(norddet):
+            wave[:, iorddet] = self[iorddet].OPT_WAVE
+            flux_gpm[:, iorddet] = self[iorddet].OPT_MASK
+            detector[iorddet] = self[iorddet].DET
+            if self[0].PYPELINE == 'Echelle':
+                ech_orders[iorddet] = self[iorddet].ECH_ORDER
+            if ret_flam:
+                flux[:, iorddet] = self[iorddet].OPT_FLAM
+                flux_ivar[:, iorddet] = self[iorddet].OPT_FLAM_IVAR
+            else:
+                flux[:, iorddet] = self[iorddet].OPT_COUNTS
+                flux_ivar[:, iorddet] = self[iorddet].OPT_COUNTS_IVAR
+
+        # Populate meta data
+        try:
+            spectrograph = load_spectrograph(self.header['PYP_SPEC'])
+        except:
+            # TODO JFH  This is a hack until a generic spectrograph is implemented.
+            spectrograph = load_spectrograph('shane_kast_blue')
+
+        meta_spec = {}
+        core_keys = spectrograph.header_cards_for_spec()
+        for key in core_keys:
+            try:
+                meta_spec[key.upper()] = self.header[key.upper()]
+            except KeyError:
+                pass
```

Comment:
I don't know the details here, but it feels like we should either do something better than these try/except blocks or actually have the code fault instead of proceeding.

Reply:
I don't like it when codes fault constantly because meta data does not exist which is not so crucial for execution. That is what is happening here.  I've added that the code will print a warning if the KeyError occurs so the user knows to take a look.